text/0000-overload-logic-operators.md

```diff
@@ -0,0 +1,228 @@
+
+    /// Complete the *logical and* in case it did not short-circuit.
+    fn logical_and(lhs: Intermediate, rhs: Rhs) -> Self::Output;
+}
+```
+
+With a matching desugaring:
+
+```rust
+<expr_a::<T>> || <expr_b::<T>>
+
+==>
+
+match expr_a.short_circuit_or() {
+    ShortCircuit::Short(res) => res,
+    ShortCircuit::Long(lhs) => logical_or(lhs, expr_b)
+}
+```
+
+and
+
+```rust
+<expr_a::<T>> && <expr_b::<T>>
+
+==>
+
+match expr_a.short_circuit_and() {
+    ShortCircuit::Short(res) => res,
+    ShortCircuit::Long(lhs) => logical_and(lhs, expr_b)
+}
+```
+
+From taking into consideration the current functions, and previous discussion on [internals](https://internals.rust-lang.org/t/pre-rfc-overload-short-curcuits/10460) it seems that the following makes the most sense in terms of outcomes.
+
+#### For `Option<T>`:
+
+```rust
+fn foo() -> Option<i32> {
+    Some(3)
+}
+
+fn main() {
+    Some(4) || Some(5); // == Some(4)
+    None || Some(5);    // == Some(5)
+    Some(4) || foo();   // == Some(4) (foo is *not* called)
+    None || foo();      // == Some(3) (foo is called)
+    None || 3;          // == 3
+    Some(2) || 1;       // == 2
+    Some(1) || panic!() // == Some(1)               + These two are side effects from !
+    None || return      // returns from function    + and are the same to how boolean || works
+    Some(2) && Some(3)  // Some(3)
+    None && Some(1)     // None
+    Some(3) && None     // None
+
+    Some(2) || Some("hello") // Error: LogicalOr<Option<&str>> not implemented for Option<i32>
+    Some(2) || 2 || 3        // Error: LogicalOr<i32> is not implemented for i32
+}
+```
+
+#### For `Result<T, E>`
+```rust
+struct MyError;
+
+fn foo() -> Result<i32, MyError> {
+    Ok(3)
+}
+
+fn main() {
+    Ok(4) || Ok(5);           // == Ok(4)
+    Err(MyError) || Ok(5);  // == Ok(5)
+    Ok(4) || foo();           // == Ok(4) (foo is *not* called)
+    Err(MyError) || foo();  // == Ok(3) (foo is called)
+    Err(MyError) || 3;      // == 3
+    Ok(2) || 1;               // == 2
+
+    Ok(2) || Ok("hello");  // Error: LogicalOr<Result<&str, _>> not implemented for Result<i32, _>
+    Ok(2) || 2 || 3;       // Error: LogicalOr<i32> is not implemented for i32
+}
+```
+
+The feature should be thought about as moving the logic from methods and into the current function.
+It maps very seamlessly from using the methods and is equivalent in use without having to worry about
+the naming convention of the short circuit methods. The same mental state of short circuit from
+bools applies directly, without having any recourse to "truthiness" which is not a desirable trait.
+
+This RFC also proposes to deprecate the `.or(...)`, `.or_with(|| ...)`, `.and(...)`, `.and_with(||
+...)`, `.unwrap_or(...)`, and `.unwrap_or_with(|| ...)` methods on `Option<T>` and `.or(...)`,
+`.or_with(|| ...)`, `.unwrap_or(...)`, and `.unwrap_or_with(|| ...)` methods on `Result<T, E>` since
+using this feature renders them unneeded.
```

Comment:
I think it may be a good idea to allow operator overloading but I don't think it is a nice idea to mix both `or_` and `unwrap_` functions into the same operator, it may be hard to distinguish if it is behaving like which function since it does it implicitly, it also requires the developer to know the context on the output of the function in order to understand how is `&&` and `||` behaving.

I think keeping `&&` and `||` for only `or` and `and` without those `unwrap_` might be a good idea for `Option` and `Result`. It reduces the load on the user mind to get to understand whether it behaves like `or_` or `unwrap_`, you never know, you need to always double check, changing one function signature (`Option<T>` to `T`) might break all other parts of the code using `&&` and `||`.

Prior art (bad art):

- python `**kwargs` implicitness, results in bad docs and mental load, similar to this in the sense that it does multiple stuff

```python
class HOTP(OTP):
    """
    Handler for HMAC-based OTP counters.
    """
    def __init__(self, *args: Any, initial_count: int = 0, **kwargs: Any) -> None:
        """
        :param initial_count: starting HMAC counter value, defaults to 0
        """
        self.initial_count = initial_count
        super(HOTP, self).__init__(*args, **kwargs)
```
Guess what could you put for `kwargs` without clicking on the item below.

<details><summary>solution (you need to read the source code to find this)</summary>

```python
class OTP(object):
    """
    Base class for OTP handlers.
    """
    def __init__(
        self,
        s: str,
        digits: int = 6,
        digest: Any = hashlib.sha1,
        name: Optional[str] = None,
        issuer: Optional[str] = None
    ) -> None:
        """
        :param s: secret in base32 format
        :param digits: number of integers in the OTP. Some apps expect this to be 6 digits, others support more.
        :param digest: digest function to use in the HMAC (expected to be sha1)
        :param name: account name
        :param issuer: issuer
        """
        self.digits = digits
        self.digest = digest
        self.secret = s
        self.name = name or 'Secret'
        self.issuer = issuer
```
</details>